File: gallery/src/main/java/com/qkopy/gallery/helper/ImageHelper.java
/*
 * Created by @ajithvgiri on 4/12/18 8:01 PM
 * Copyright (c) 2018 . All rights reserved.
 * Last modified 4/12/18 7:05 PM
 */

package com.qkopy.gallery.helper;

import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.content.pm.ResolveInfo;
import android.net.Uri;
import android.os.Environment;
import android.util.Log;
import com.qkopy.gallery.BuildConfig;
import com.qkopy.gallery.model.Image;
import com.qkopy.gallery.model.SavePath;

import java.io.File;
import java.io.IOException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;


public class ImageHelper {

    private static final String TAG = ImageHelper.class.getSimpleName();

    public static File createImageFile(SavePath savePath) {
        // External sdcard location
        final String path = savePath.getPath();
        File mediaStorageDir = savePath.isFullPath() ? new File(path) : new File(Environment.getExternalStoragePublicDirectory(BuildConfig.STORAGE_PATH), path);

        // Create the storage directory if it does not exist
        if (!mediaStorageDir.exists()) {
            if (!mediaStorageDir.mkdirs()) {
                Log.d(TAG, "Oops! Failed create " + path);
                return null;
            }
        }

        String timeStamp = new SimpleDateFormat("yyyyMMdd_HHmmss").format(new Date());
        String imageFileName = "IMG_" + timeStamp;

        File imageFile = null;
        try {
            imageFile = File.createTempFile(imageFileName, ".jpg", mediaStorageDir);
        } catch (IOException e) {
            Log.d(TAG, "Oops! Failed create " + imageFileName + " file");
        }
        return imageFile;
    }

    public static String getNameFromFilePath(String path) {
        if (path.contains(File.separator)) {
            return path.substring(path.lastIndexOf(File.separator) + 1);
        }
        return path;
    }

    public static void grantAppPermission(Context context, Intent intent, Uri fileUri) {
        List<ResolveInfo> resolvedIntentActivities = context.getPackageManager()
                .queryIntentActivities(intent, PackageManager.MATCH_DEFAULT_ONLY);

        for (ResolveInfo resolvedIntentInfo : resolvedIntentActivities) {
            String packageName = resolvedIntentInfo.activityInfo.packageName;
            context.grantUriPermission(packageName, fileUri,
                    Intent.FLAG_GRANT_WRITE_URI_PERMISSION | Intent.FLAG_GRANT_READ_URI_PERMISSION);
        }
    }

    public static void revokeAppPermission(Context context, Uri fileUri) {
        context.revokeUriPermission(fileUri,
                Intent.FLAG_GRANT_WRITE_URI_PERMISSION | Intent.FLAG_GRANT_READ_URI_PERMISSION);
    }

    public static List<Image> singleListFromPath(String path) {
        List<Image> images = new ArrayList<>();
        images.add(new Image(0, getNameFromFilePath(path), path,false));
        return images;
    }

    public static boolean isGifFormat(Image image) {
        String extension = image.getPath().substring(image.getPath().lastIndexOf(".") + 1, image.getPath().length());
        return extension.equalsIgnoreCase("gif");
    }


}

File: gallery/src/main/java/com/qkopy/gallery/helper/ImageHelper.kt
package com.qkopy.gallery.helper

import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Environment
import android.util.Log
import com.qkopy.gallery.BuildConfig
import com.qkopy.gallery.model.Image
import com.qkopy.gallery.model.SavePath
import java.io.File
import java.io.IOException
import java.text.SimpleDateFormat
import java.util.*

object ImageHelper {
    private val TAG = ImageHelper::class.java.simpleName
    fun createImageFile(context: Context,savePath: SavePath): File? {
        // External sdcard location
        val path: String = savePath.path

        val mediaStorageDir =
            if (savePath.isFullPath) File(path) else File(
                context.getExternalFilesDir(Environment.DIRECTORY_PICTURES),
                path
            )

        // Create the storage directory if it does not exist
        if (!mediaStorageDir.exists()) {
            if (!mediaStorageDir.mkdirs()) {
                Log.d(TAG, "Oops! Failed create $path")
                return null
            }
        }
        val timeStamp =
            SimpleDateFormat("yyyyMMdd_HHmmss").format(Date())
        val imageFileName = "IMG_$timeStamp"
        var imageFile: File? = null
        try {
            imageFile = File.createTempFile(imageFileName, ".jpg", mediaStorageDir)
        } catch (e: IOException) {
            Log.d(TAG, "Oops! Failed create $imageFileName file")
        }
        return imageFile
    }

    fun getNameFromFilePath(path: String): String {
        return if (path.contains(File.separator)) {
            path.substring(path.lastIndexOf(File.separator) + 1)
        } else path
    }

    fun grantAppPermission(
        context: Context,
        intent: Intent?,
        fileUri: Uri?
    ) {
        val resolvedIntentActivities =
            context.packageManager
                .queryIntentActivities(intent, PackageManager.MATCH_DEFAULT_ONLY)
        for (resolvedIntentInfo in resolvedIntentActivities) {
            val packageName = resolvedIntentInfo.activityInfo.packageName
            context.grantUriPermission(
                packageName, fileUri,
                Intent.FLAG_GRANT_WRITE_URI_PERMISSION or Intent.FLAG_GRANT_READ_URI_PERMISSION
            )
        }
    }

    fun revokeAppPermission(
        context: Context,
        fileUri: Uri?
    ) {
        context.revokeUriPermission(
            fileUri,
            Intent.FLAG_GRANT_WRITE_URI_PERMISSION or Intent.FLAG_GRANT_READ_URI_PERMISSION
        )
    }

    fun singleListFromPath(path: String): List<Image> {
        val images: MutableList<Image> = ArrayList<Image>()
        images.add(Image(0, getNameFromFilePath(path), path, false))
        return images
    }

    fun isGifFormat(image: Image): Boolean {
        val extension: String = image.path
            .substring(image.path.lastIndexOf(".") + 1, image.path.length)
        return extension.equals("gif", ignoreCase = true)
    }
}
